src/azure-cli-core/azure/cli/core/__init__.py

```diff
@@ -5,7 +5,143 @@
+import os
+
+from knack.cli import CLI
+from knack.commands import CLICommandsLoader, CLICommand
+from knack.introspection import extract_args_from_signature, extract_full_summary_from_signature
+from knack.log import get_logger
+import six
+
+ARGCOMPLETE_ENV_NAME = '_ARGCOMPLETE'
+
+logger = get_logger(__name__)
+
+class AzCli(CLI):
+
+    def __init__(self, **kwargs):
+        super(AzCli, self).__init__(**kwargs)
+        self.session =  {
+            'headers': {},  # the x-ms-client-request-id is generated before a command is to execute
+            'command': 'unknown',
+            'completer_active': ARGCOMPLETE_ENV_NAME in os.environ,
+            'query_active': False
+        }
+
+    def get_cli_version(output):
+        from azure.cli.core.util import show_version_info_exit
+        return show_version_info_exit(None)
+
+
+class AzCliCommand(CLICommand):
+
+    def _resolve_default_value_from_cfg_file(self, arg, overrides):
+        from azure.cli.core._config import DEFAULTS_SECTION
+
+        if not hasattr(arg.type, 'required_tooling'):
+            required = arg.type.settings.get('required', False)
+            setattr(arg.type, 'required_tooling', required)
+        if 'configured_default' in overrides.settings:
+            def_config = overrides.settings.pop('configured_default', None)
+            setattr(arg.type, 'default_name_tooling', def_config)
+            # same blunt mechanism like we handled id-parts, for create command, no name default
+            if (self.name.split()[-1] == 'create' and overrides.settings.get('metavar', None) == 'NAME'):
+                return
+            setattr(arg.type, 'configured_default_applied', True)
+            print('Context = {}', type(self.ctx))
+            config_value = self.ctx.config.get(DEFAULTS_SECTION, def_config, None)
+            if config_value:
+                logger.warning("Using default '%s' for arg %s", config_value, arg.name)
+                overrides.settings['default'] = config_value
+                overrides.settings['required'] = False
+
+    def update_argument(self, param_name, argtype):
+        arg = self.arguments[param_name]
+        self._resolve_default_value_from_cfg_file(arg, argtype)
+        arg.type.update(other=argtype)
+
+
+class MainCommandsLoader(CLICommandsLoader):
+
+    def __init__(self, ctx=None):
+        super(MainCommandsLoader, self).__init__(ctx)
+        from azure.cli.command_modules.redis import RedisCommandsLoader
+        from azure.cli.command_modules.profile import ProfileCommandsLoader
+        from azure.cli.command_modules.cloud import CloudCommandsLoader
+        from azure.cli.command_modules.feedback import FeedbackCommandsLoader
+        self.loaders = [
+            RedisCommandsLoader(ctx=self.ctx),
+            ProfileCommandsLoader(ctx=self.ctx),
+            CloudCommandsLoader(ctx=self.ctx),
+            FeedbackCommandsLoader(ctx=self.ctx)
+        ]
```

Comment:
Would have to look at automatically detecting the modules.
It may not have to differ too far from https://github.com/Azure/azure-cli/blob/master/src/azure-cli-core/azure/cli/core/commands/__init__.py#L375.